### PR TITLE
Implement TopP, TopK and Beam samplers

### PR DESCRIPTION
**Pull request by chenmoneygithub**
A few things covered:
- Implement TopP, TopK and Beam samplers
- Add "Sampler" suffix to our sampler class names.
- Add `run_eagerly` option to our sampler to align with `model.compile()`.

One special thing to note:
- In Beam sampler, there is a strange error on `batch_size=1` case, which is after the first iteration, the shape of `beam_probs` changes from `[1, None]` to `[None, None]`, so we add `shape_invariants` specifically for beam sampler.

### CODE REVIEW

File: keras_nlp/samplers/__init__.py
Comment (jbischof) [inline keras_nlp/samplers/__init__.py:17]: Clever!

File: keras_nlp/samplers/beam_sampler.py
Comment (jbischof) [inline keras_nlp/samplers/beam_sampler.py:33]: implements the

File: keras_nlp/samplers/beam_sampler.py
Comment (jbischof) [inline keras_nlp/samplers/beam_sampler.py:48]: @chenmoneygithub I really don't think these arg blocks match the rest of the library's style. I don't think there's a perfect answer but I'd prefer it not to be obvious who wrote what 🥗

File: keras_nlp/samplers/sampler.py
Comment (jbischof) [inline keras_nlp/samplers/sampler.py:150]: Do we need two flags or just one then? What happens if they are both `False`---what is a "non-XLA" graph? I'm confused!

**Review by jbischof — CHANGES_REQUESTED** [review]
Thanks @chenmoneygithub looking great! The biggest thing is unifying our docstring style

File: keras_nlp/samplers/sampler.py
Comment (chenmoneygithub) [reply keras_nlp/samplers/sampler.py:150]: This is following the style of `model.compile()`: [code link](https://github.com/keras-team/keras/blob/e6784e4302c7b8cd116b74a784f4b78d60e83c26/keras/engine/training.py#L756)

non-XLA graph is quite common, anything annotated with `tf.function` without `jit_compile=True` is a non-xla graph.

File: keras_nlp/samplers/beam_sampler.py
Comment (chenmoneygithub) [reply keras_nlp/samplers/beam_sampler.py:48]: Thanks! I checked our code base, it appears we have a few code having this style:
- maskedLmhead: [link](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/masked_lm_head.py#L71)
- PositionEmbedding: [link](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/position_embedding.py#L54)
- SinePositionEmbedding: [link](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/layers/position_embedding.py#L54)
- Generation utils: [link](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/utils/text_generation.py#L131)

I feel like giving these numbers an explicit meaning makes the code longer but kinda easier to understand.

Comment (jbischof) [reply keras_nlp/samplers/beam_sampler.py:48]: It's a totally valid opinion @chenmoneygithub but it's also important for the code to have a unified style rather than each contributor producing different looking code. If there's an example that's unclear without named params I'm open to trying something different but otherwise I'm hoping we can compromise!

File: keras_nlp/samplers/sampler.py
Comment (jbischof) [reply keras_nlp/samplers/sampler.py:150]: Yes, agree, was just trying to understand the difference

File: keras_nlp/samplers/beam_sampler.py
Comment (mattdangerw) [reply keras_nlp/samplers/beam_sampler.py:48]: Taking steps towards a more unified style (and then reflecting that in our style guide) sgtm. What are the main places this differs, besides the constants at the top?

Comment (jbischof) [reply keras_nlp/samplers/beam_sampler.py:48]: Basically what I asked for in https://github.com/keras-team/keras-nlp/issues/658 was the current thought. A bit less script-like and a little more "drop this line in colab and see what we're talking about".

File: keras_nlp/samplers/beam_sampler.py
Comment (mattdangerw) [inline keras_nlp/samplers/beam_sampler.py:213]: return immediately

File: keras_nlp/samplers/top_p_sampler.py
Comment (mattdangerw) [inline keras_nlp/samplers/top_p_sampler.py:109]: The main change I could see us making if we want to, is splitting the overload points into `sample` and `sample_step` (I think some discussion of this on other PR?).

Overload `sample` if you want to control the whole process (e.g. beam search). Overload `sample_step` if you want to control simply going from one probability distribution to one sample. E.g. the body of this class could look like:

```python
def sample_step(preds):
     sorted_preds, sorted_indices = tf.math.top_k(
          pred, k=tf.shape(preds)[1], sorted=True
      )
      # Calculate cumulative probability distribution.
      cumulative_probs = tf.math.cumsum(sorted_preds, axis=-1)
      # Create a mask for the tokens to keep.
      keep_mask = cumulative_probs <= self.p
      # Shift to include the last token that exceed p.
      shifted_keep_mask = tf.concat(
          [tf.ones_like(keep_mask[:, :1]), keep_mask[:, :-1]], axis=-1
      )
      # Filter out unmasked tokens and sample from filtered distribution.
      probs = tf.where(
          shifted_keep_mask,
          sorted_preds,
          tf.zeros(tf.shape(pred), dtype=sorted_preds.dtype),
      )
      sorted_next_token = tf.random.categorical(
          tf.math.log(probs), 1, seed=self.seed
      )
      return tf.gather_nd(
          sorted_indices, sorted_next_token, batch_dims=1
      )
```

This might improve the readability of our simple samples, while still keeping full extensibility.

File: keras_nlp/samplers/beam_sampler.py
Comment (mattdangerw) [reply keras_nlp/samplers/beam_sampler.py:48]: I could see us replacing the "model" with something like `tf.random.uniform(shape, minval=-1, maxval=1)`. It is kind of weird to me that we show a whole model, that is trainable but randomly initialized (so results will be random anyway), and not even sequence aware so would never really perform even if your trained it. For a new user this seems a bit of a red herring.

Would be more concise to do something like:
```python
    def token_probability_fn(inputs, mask):
        return tf.random.uniform(...) # Replace with a real model!
```

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Thanks Chen! Left some comments and a few questions.

File: keras_nlp/samplers/beam_sampler.py
Comment (chenmoneygithub) [reply keras_nlp/samplers/beam_sampler.py:48]: I would like to keep the model part so that the example is closer to real use cases. 

I will unify the docstring to move those hypers inline.

File: keras_nlp/samplers/top_p_sampler.py
Comment (chenmoneygithub) [reply keras_nlp/samplers/top_p_sampler.py:96]: yea, I should delete the docstring formatting. Done.

File: keras_nlp/samplers/top_p_sampler.py
Comment (chenmoneygithub) [reply keras_nlp/samplers/top_p_sampler.py:109]: As long as we can make `BeamSampler` an outlier, I am down to this refactoring!

Actually we can move one step further to only leave `get_next_token` open, which takes in a prob distribution and returns the next token, and the rest updating logic can be shared across those samplers. 

Reflected this change in the PR.

Comment (jbischof) [reply keras_nlp/samplers/top_p_sampler.py]: Yes, this is what I was hoping for! 🚀

File: keras_nlp/samplers/greedy_sampler.py
Comment (jbischof) [inline keras_nlp/samplers/greedy_sampler.py:43]: Let's inline this arg as well. Or am I missing something?

File: keras_nlp/samplers/sampler.py
Comment (jbischof) [inline keras_nlp/samplers/sampler.py:250]: Do you think sliding window can fit into this paradigm?

File: keras_nlp/samplers/greedy_sampler.py
Comment (jbischof) [inline keras_nlp/samplers/greedy_sampler.py:66]: For clarity you could use named args:

```python
print(sampler(prompt, token_probability_fn, max_length=10))
```

This matches or perhaps enhances the readability of using globals.

**Review by jbischof — APPROVED** [review]
Just some nits. Looking good!

File: keras_nlp/samplers/greedy_sampler.py
Comment (chenmoneygithub) [reply keras_nlp/samplers/greedy_sampler.py:43]: This one is used twice - embedding and dense, and the arg name does not suggest "vocab_size", so I am keeping this one for clarify.

File: keras_nlp/samplers/top_k_sampler.py
Comment (mattdangerw) [inline keras_nlp/samplers/top_k_sampler.py:95]: Don't you need a `get_config()` adding the k param?

File: keras_nlp/samplers/sampler.py
Comment (chenmoneygithub) [reply keras_nlp/samplers/sampler.py:250]: yea, it should work

File: keras_nlp/samplers/top_p_sampler.py
Comment (mattdangerw) [inline keras_nlp/samplers/top_p_sampler.py:108]: Same here add `get_config`

File: keras_nlp/samplers/top_k_sampler.py
Comment (mattdangerw) [inline keras_nlp/samplers/top_k_sampler.py:95]: Don't you need a `get_config()` adding the k param?

File: keras_nlp/samplers/greedy_sampler.py
Comment (chenmoneygithub) [reply keras_nlp/samplers/greedy_sampler.py:66]: good call, done

**Review by mattdangerw — APPROVED** [review]
Left a few more comments re `get_config`. +approval, as I am fine to merge after changes to get something landed. But hope we will stay open to changes here as we dig more into the generative case!

**Comment by jbischof** [comment]
Wait for @fchollet, he wants to take a look and can make the final call on class names.

**Comment by chenmoneygithub** [comment]
Let's merge this one to unblock the TFLite sprint, I will sync with Francois offline regarding the string identifier.

File: keras_nlp/samplers/beam_sampler.py
Comment (fchollet) [inline keras_nlp/samplers/beam_sampler.py:85]: This should raise an error I suppose?

**Pull request merged**
